vochain/cometbft.go

```diff
@@ -499,28 +503,140 @@ func (app *BaseApplication) ProcessProposal(_ context.Context,
 	}, nil
 }
 
-// ListSnapshots returns a list of available snapshots.
-func (*BaseApplication) ListSnapshots(context.Context,
-	*cometabcitypes.ListSnapshotsRequest) (*cometabcitypes.ListSnapshotsResponse, error) {
-	return &cometabcitypes.ListSnapshotsResponse{}, nil
+// Example StateSync (Snapshot) successful flow:
+// Alice is a comet node, up-to-date with RPC open on port 26657
+// Bob is a fresh node that is bootstrapping, with params:
+//    * StateSync.Enable = true
+//    * StateSync.RPCServers = alice:26657, alice:26657
+// * Bob comet will ask Alice for snapshots (app doesn't intervene here)
+// * Alice comet calls ListSnapshots, App returns a []*v1.Snapshot
+// * Bob comet calls OfferSnapshot passing a single *v1.Snapshot, App returns ACCEPT
+// * Bob comet asks Alice for chunks (app doesn't intervene here)
+// * Alice comet calls (N times) LoadSnapshotChunk passing height, format, chunk index. App returns []byte
+// * Bob comet calls (N times) ApplySnapshotChunk passing []byte, chunk index and sender. App returns ACCEPT
+
+// ListSnapshots provides cometbft with a list of available snapshots.
+func (app *BaseApplication) ListSnapshots(_ context.Context,
+	req *cometabcitypes.ListSnapshotsRequest,
+) (*cometabcitypes.ListSnapshotsResponse, error) {
+	list := app.State.ListSnapshots()
+
+	response := &cometabcitypes.ListSnapshotsResponse{}
+	for height, dsi := range list {
+		response.Snapshots = append(response.Snapshots, &cometabcitypes.Snapshot{
+			Height:   uint64(height),
+			Format:   0,
+			Chunks:   dsi.CalculateChunks(app.stateSyncChunkSize),
+			Hash:     dsi.Hash[:],
+			Metadata: []byte{},
+		})
+	}
+	log.Debugf("CometBFT requests our list of Snapshots, we offer %d options", len(response.Snapshots))
+	return response, nil
 }
 
-// OfferSnapshot returns the response to a snapshot offer.
-func (*BaseApplication) OfferSnapshot(context.Context,
-	*cometabcitypes.OfferSnapshotRequest) (*cometabcitypes.OfferSnapshotResponse, error) {
-	return &cometabcitypes.OfferSnapshotResponse{}, nil
+// horrible hack, turn this into a field of app or state package, and use something concurrent-safe
+var snaps = struct {
+	Snap           *cometabci.Snapshot
+	ChunkFilenames map[uint32]string
+}{}
+
+// OfferSnapshot is called by cometbft during StateSync, when another node offers a Snapshot.
+func (*BaseApplication) OfferSnapshot(_ context.Context,
+	req *cometabcitypes.OfferSnapshotRequest,
+) (*cometabcitypes.OfferSnapshotResponse, error) {
+	log.Debugw("CometBFT offers us a Snapshot",
```

Comment:
We don't use capital letters on log messages

Reply:
47dc8891